pkg/scheduler/cache/event_handlers.go

```diff
@@ -402,6 +405,69 @@ func (sc *SchedulerCache) DeleteNode(obj interface{}) {
 	}
 }
 
+func (sc *SchedulerCache) AddOrUpdateCSINode(obj interface{}) {
+	csiNode, ok := obj.(*sv1.CSINode)
+	if !ok {
+		return
+	}
+
+	var csiNodeStatus *schedulingapi.CSINodeStatusInfo
+	var found bool
+	sc.Mutex.Lock()
+	defer sc.Mutex.Unlock()
+	// update nodeVolumeCount
+
+	if csiNodeStatus, found = sc.CSINodesStatus[csiNode.Name]; !found {
+		csiNodeStatus = &schedulingapi.CSINodeStatusInfo{
+			CSINodeName:  csiNode.Name,
+			DriverStatus: make(map[string]bool),
+		}
+		sc.CSINodesStatus[csiNode.Name] = csiNodeStatus
+	}
+
+	for i := range csiNode.Spec.Drivers {
+		d := csiNode.Spec.Drivers[i]
+		csiNodeStatus.DriverStatus[d.Name] = d.Allocatable != nil && d.Allocatable.Count != nil
+	}
+}
+
+func (sc *SchedulerCache) UpdateCSINode(oldObj, newObj interface{}) {
+	oldCSINode, ok := oldObj.(*sv1.CSINode)
+	if !ok {
+		return
+	}
+	newCSINode, ok := newObj.(*sv1.CSINode)
+	if !ok {
+		return
+	}
+	if reflect.DeepEqual(oldCSINode.Spec, newCSINode.Spec) {
+		return
+	}
+	sc.AddOrUpdateCSINode(newObj)
+}
+
+func (sc *SchedulerCache) DeleteCSINode(obj interface{}) {
+	var csiNode *sv1.CSINode
+	switch t := obj.(type) {
+	case *sv1.CSINode:
+		csiNode = obj.(*sv1.CSINode)
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		csiNode, ok = t.Obj.(*sv1.CSINode)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("unable to convert object %T to *sv1.CSINode", obj))
```

Comment:
what's the difference between klog and utilruntime.HandleError? if they are the same, we can use the unified format klog and do not have to introcue utiruntime package.

Reply:
address comments